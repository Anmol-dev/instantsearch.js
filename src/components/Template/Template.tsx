/** @jsx h */

import { createElement } from 'preact';
import { memo } from 'preact/compat';
import { renderTemplate, isEqual } from '../../lib/utils';
import { Template as WidgetTemplate } from '../../types';

export type TemplateProps = {
  templateKey: string;
  templates: { [templateKey: string]: WidgetTemplate<any> };
  rootTagName?: string;
  rootProps?: object;
  data?: any;
};

function Template({
  data = {},
  templateKey,
  rootProps,
  rootTagName = 'div',
  templates,
}: TemplateProps) {
  const content: string = renderTemplate({
    templates,
    templateKey,
    data,
    helpers: {},
  });

  return createElement(rootTagName, {
    ...rootProps,
    dangerouslySetInnerHTML: { __html: content },
  });
}

function areEqual(prevProps: TemplateProps, nextProps: TemplateProps) {
  return (
    prevProps.templateKey === nextProps.templateKey &&
    isEqual(prevProps.data, nextProps.data) &&
    isEqual(prevProps.rootProps, nextProps.rootProps)
  );
}

export default memo(Template, areEqual);
