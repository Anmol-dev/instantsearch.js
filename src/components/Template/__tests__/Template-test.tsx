/** @jsx h */

import { h } from 'preact';
import Template, { TemplateProps } from '../Template';
import { render } from '@testing-library/preact';

type TemplateData = {
  items: string[];
};

const defaultProps: TemplateProps<TemplateData> = {
  template: 'Template content',
  data: {
    items: [],
  },
};

describe('Template', () => {
  it('can configure custom rootTagName', () => {
    const props = {
      ...defaultProps,
      rootTagName: 'span',
    };

    const { container } = render(<Template {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('forward rootProps to the first node', () => {
    function onClick() {}
    const props = {
      ...defaultProps,
      rootProps: { className: 'className', onClick },
    };

    const { container } = render(<Template {...props} />);

    expect(container).toMatchSnapshot();
  });

  describe('memo', () => {
    let renderCount: number;

    beforeEach(() => {
      renderCount = 0;
    });

    it('does not render when `data` does not change', () => {
      const props = {
        ...defaultProps,
        template: () => {
          renderCount++;
          return 'test';
        },
      };

      const { rerender } = render(<Template {...props} />);
      rerender(<Template {...props} />);

      expect(renderCount).toEqual(1);
    });

    it('renders when `data` changes', () => {
      const props = {
        ...defaultProps,
        template: () => {
          renderCount++;
          return 'test';
        },
      };

      const { rerender } = render(<Template {...props} />);

      expect(renderCount).toEqual(1);

      const newProps = {
        ...props,
        data: { items: ['1'] },
      };

      rerender(<Template {...newProps} />);

      expect(renderCount).toEqual(2);
    });

    it('renders when `rootProps` changes', () => {
      const props = {
        ...defaultProps,
        template: () => {
          renderCount++;
          return 'test';
        },
      };

      const { rerender } = render(<Template {...props} />);

      expect(renderCount).toEqual(1);

      const newProps = {
        ...props,
        rootProps: {
          className: 'className',
        },
      };

      rerender(<Template {...newProps} />);

      expect(renderCount).toEqual(2);
    });

    it('does not render when `rootProps` does not change', () => {
      const props = {
        ...defaultProps,
        rootProps: {
          className: 'initialClassName',
        },
        template: () => {
          renderCount++;
          return 'test';
        },
      };

      const { rerender } = render(<Template {...props} />);

      expect(renderCount).toEqual(1);

      const newProps = {
        ...props,
        rootProps: {
          className: 'initialClassName',
        },
      };

      rerender(<Template {...newProps} />);

      expect(renderCount).toEqual(1);
    });
  });
});
