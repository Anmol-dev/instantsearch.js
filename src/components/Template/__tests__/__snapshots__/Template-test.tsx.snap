// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template can configure custom rootTagName 1`] = `
<div>
  <span>
    Template content
  </span>
</div>
`;

exports[`Template forward rootProps to the first node 1`] = `
<div>
  <div
    class="className"
  >
    Template content
  </div>
</div>
`;
